token/program-2022-test/tests/memo_transfer.rs

```diff
@@ -38,8 +54,62 @@ async fn test_memo_transfers(
     assert!(bool::from(extension.require_incoming_transfer_memos));
 
     // attempt to transfer from alice to bob without memo
-    // TODO: should fail when token/program-2022/src/processor.rs#L376 is completed
+    let err = token
+        .transfer_unchecked(&alice_account, &bob_account, &alice, 10)
+        .await
+        .unwrap_err();
+    assert_eq!(
+        err,
+        TokenClientError::Client(Box::new(TransportError::TransactionError(
+            TransactionError::InstructionError(
+                0,
+                InstructionError::Custom(TokenError::NoMemo as u32)
+            )
+        )))
+    );
+    let bob_state = token.get_account_info(&bob_account).await.unwrap();
+    assert_eq!(bob_state.base.amount, 0);
+
+    // attempt to transfer from alice to bob with misplaced memo
+    let mut ctx = context.lock().await;
+    #[allow(deprecated)]
+    let instructions = vec![
+        spl_memo::build_memo(&[240, 159, 166, 150], &[]),
+        system_instruction::transfer(&ctx.payer.pubkey(), &alice.pubkey(), 42),
+        spl_token_2022::instruction::transfer(
+            &spl_token_2022::id(),
+            &alice_account,
+            &bob_account,
+            &alice.pubkey(),
+            &[],
+            10,
+        )
+        .unwrap(),
+    ];
+    let tx = Transaction::new_signed_with_payer(
+        &instructions,
+        Some(&ctx.payer.pubkey()),
+        &[&ctx.payer, &alice],
+        ctx.last_blockhash,
+    );
+    let err: TransactionError = ctx
+        .banks_client
+        .process_transaction(tx)
+        .await
+        .unwrap_err()
+        .unwrap()
+        .into();
+    drop(ctx);
+    assert_eq!(
+        err,
+        TransactionError::InstructionError(2, InstructionError::Custom(TokenError::NoMemo as u32))
+    );
+    let bob_state = token.get_account_info(&bob_account).await.unwrap();
+    assert_eq!(bob_state.base.amount, 0);
+
+    // transfer with memo
```

Comment:
Feel free to ignore this one, since I don't know too much about the evolution of memo, but is it worth also adding a test that explicitly uses memo v1?

Reply:
Definitely wouldn't hurt! Memo v1 still has a fair bit of use, and is actually a bit cheaper compute-wise, since it doesn't log the memo.